### PR TITLE
Add addItems message-flow and functionality to Cart's to fix #48

### LINKED ISSUE #48: [sampleapp] Add `add_items` messaging and logic to `Cart`
# Issue

This is an issue w.r.t. the **sample application**.
Add `add_items` messaging and logic to `Cart`.

# Problem Description

- `Cart` is missing the top-level messaging and logic for `add_items`

# Supporting Information

- Refer to the manifesto paper for reference

### PR DESCRIPTION
**Pull request by srfc**
Fixes #48

## Proposed Changes

  - add `add_items` logic to `Cart` dactor
  - first multiple actor spanning procedure

## Related

  - #49, #50
  - depends on #58 (**merge first!**)

### CODE REVIEW

**Comment by srfc** [comment]
This ended up being less straight forward than I thought.

Currently I am using the `akka` `ask` pattern to keep track of responses. Since this is quite obviously in contradiction to

> Don't ask, tell.

I'm open to exploring alternatives. For now we should focus on getting the example running though, I think.

File: sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala [hunk 1]
```diff
@@ -54,8 +65,123 @@ class Cart(id: Int) extends Dactor(id) {
...
 
+  def addItems(orders: Seq[AddItems.Order], customerId: Int): Future[Try[Seq[Record]]] = {
+    // ask for prices of orders
+    // i_id, i_price, i_min_price
+    val prices = Future.sequence(orders
+      .groupBy(order => order.sectionId)
+      .map(askStoreSectionForPrices))
+
+    // c_g_id
+    val customer_group_id = askCustomerForGroupId(customerId)
```
Comment (CodeLionX) [inline sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala]: Why are you mixing _camel_case_ and names concatenated with underscore?

File: sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala [hunk 2]
```diff
@@ -54,8 +65,123 @@ class Cart(id: Int) extends Dactor(id) {
...
+    val i_ids_and_group_id = for {
+      prices <- prices
+      c_g_id <- customer_group_id
+    } yield (prices.flatten.map(ppr => ppr.i_id), c_g_id)
+
+    // i_id, fixed_disc
+    val fixed_disc = i_ids_and_group_id.flatMap(tuple => {
+      val (i_ids, group_id) = tuple
+      askGroupManagerForDiscount(i_ids.toSeq, group_id)
+    })
```
Comment (CodeLionX) [inline sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala]: **Style - this is subjective**

There is an alternative syntax for this, you may consider:

```scala
val fixed_disc = i_ids_and_group_id.flatMap{
  case (i_ids, group_id) => askGroupManagerForDiscount(i_ids.toSeq, group_id)
}
```

File: sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala [hunk 3]
```diff
@@ -54,8 +65,123 @@ class Cart(id: Int) extends Dactor(id) {
...
+          .withCellContent(CartPurchases.minPrice)(i_min_price)
+          .withCellContent(CartPurchases.fixedDiscount)(i_fixed_disc)
+          .build()
+    }
+
+  case class PriceDiscountPartialResult(s_id: Int, i_id: Int, i_price: Double, i_min_price: Double, i_fixed_disc: Double)
+  private def joinPriceAndDisc(ppr: Seq[Cart.this.PricePartialResult], dpr: Seq[Cart.this.DiscountsPartialResult]) = {
+    val discount_for_i_id = (dpr map {
+      a => a.i_id -> a.i_fixed_disc
+    }).toMap
```
Comment (CodeLionX) [inline sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala]: **Style - this is subjective**

Use dot-notation for getting rid of parenthesis:

```scala
val discount_for_i_id = dpr.map { a =>
  a.i_id -> a.i_fixed_disc
}.toMap
```

File: sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala [hunk 4]
```diff
@@ -54,8 +65,123 @@ class Cart(id: Int) extends Dactor(id) {
...
+          .withCellContent(CartPurchases.sessionId)(currentSessionId)
+          .withCellContent(CartPurchases.inventoryId)(i_id)
+          .withCellContent(CartPurchases.quantity)(quantityOfInventoryId.getOrElse(i_id, 0))
+          .withCellContent(CartPurchases.price)(i_price)
+          .withCellContent(CartPurchases.minPrice)(i_min_price)
+          .withCellContent(CartPurchases.fixedDiscount)(i_fixed_disc)
+          .build()
+    }
+
+  case class PriceDiscountPartialResult(s_id: Int, i_id: Int, i_price: Double, i_min_price: Double, i_fixed_disc: Double)
```
[the remark this answers is not in the document]
Comment (CodeLionX) [reply sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala]: You may even consider adding function defs to them, eg.:
```scala
case class PriceDiscountPartialResult(s_id: Int, i_id: Int, i_price: Double, i_min_price: Double, i_fixed_disc: Double) {
  def toDiscountMap: Map[Int, Double] = Map(s_id -> i_fxed_disc)
}
```

and then in L132:
```scala
val discount_for_i_id = dpr.map(_.toDiscountMap)
```

[the remark this answers is not in the document]
Comment (srfc) [reply sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala]: But `dpr` is a `Seq[PriceDiscountPartialResult`. Given your `toDiscount` function I will still have to concatenate a `Seq[Map[Int, Double]]` giving me:

```scala
      val discountForInventoryId = dpr.map(_.toDiscountMap).foldLeft(Map[Int, Double]()) { (existing, n) => existing ++ n}
```

Which I don't think is much nicer than is is now?! I can change it if you think it's clearer but right now I don't see the advantage.

File: sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Customer.scala [hunk 5]
```diff
@@ -89,6 +103,11 @@ class Customer(id: Int) extends Dactor(id) {
 
   def getCustomerInfo(): Try[Seq[Record]] = CustomerInfo.records
 
+  def getCustomerGroupId(): Try[Int] = {
+    assert(CustomerInfo.records.get.size == 1)
```
[the remark this answers is not in the document]
Comment (srfc) [reply sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Customer.scala]: Yeah, we should really have a `SingleRowRelation` for cases like this one. What do you think?

**Comment by srfc** [comment]
@CodeLionX I incorporated the changes we have talked about - can you have another look around?

**Comment by CodeLionX** [comment]
merged #58 and fixed error with relational store in companion object with this.

**Pull request merged**